### DISCUSSION TITLE
Trying to deploy my Nuxt 3 app to Firebase

### DISCUSSION THREAD

**Discussion by tavishigupta** [nuxt/nuxt#16506] [Questions]
I followed https://v3.nuxtjs.org/docs/deployment/firebase/ pretty closely.

The only differences in my code:
1. In my .firebaserc I have 2 projects instead of 1, and none of them are default (staging & prod) - I doubt this will be an issue
2. In my firebase.json I do not have the site attribute as I use `firebase use <project-alias>` to set the project.

I enter this command (as per the documentation):
`NITRO_PRESET=firebase yarn build` 
`firebase deploy`

I get this error:
`Error: could not deploy functions because the ".output/server" directory was not found. Please create it or specify a different source directory in firebase.json`

It seems like I need to do something with my firebase functions, but I just don't know what. Any help is appreciated!

**Comment by 101msk** [comment]
".output/server" directory was not found.

Would you like to do `NITRO_PRESET=firebase npm run build`?

**Comment by PMLS3** [comment]
I'm getting the following: NITRO_PRESET=firebase : The term 'NITRO_PRESET=firebase' is not recognized as the name of a cmdlet

**Reply by dosstx** [reply]
same here. ANyone got a successful deploy?

**Reply by PMLS3** [reply]
you can add this in your script, works for me "deploy-hosting": "npx cross-env NITRO_PRESET=firebase yarn build && firebase deploy"

**Comment by raydenchan** [comment]
I have similar problem too, when deploy to firebase, I get this error message:

Error: There was an error reading .output\server\package.json:
 .output\server\index.js does not exist, can't deploy Cloud Functions

**Comment by raydenchan** [comment]
Has anyone successful deployed Nuxt3 to Firebase? 
This link is down: https://v3.nuxtjs.org/guide/deployment/firebase/
Do I need Firebase Function in order to deploy to Firebase? 
I am trying to build static website with Nuxt, please guide, thank you.

**Reply by Apricitine** [reply]
I was able to get it deployed but it took a lot of tinkering and debugging. The process could definitely be more streamlined. For reference, take a look at this repo: https://github.com/Apricitine/apricite

First, make sure that the following are all installed as dev dependencies:
```
firebase-admin
firebase-functions
firebase-functions-test
```
You'll also need the `firebase` package installed too, of course.

The next couple of steps would take a while to explain. Refer to the above link and tell me if you need assistance. :)

**Reply by davidstackio** [reply]
For those on windows, you can create a `deploy.bat` file to help with deployment commands:

```
NITRO_PRESET=firebase
npm run build
cd .output/server
npm install
firebase deploy
cd ../..
```

This works if you have `firebase-tools` already installed globally.

**EDIT**
Alternatively, you can just add a [`predeploy`](https://firebase.google.com/docs/hosting/test-preview-deploy#hooks) script to your `firebase.json` instead of creating a shell script:

```
"hosting": {
    "predeploy": [
      "npm run build && cd .output/server && npm install"
    ],
    ...
```

Just make sure you have a `nitro.config.ts` file in your root directory with this (so you don't need the `NITRO_PRESET=firebase` bit):
```
import { defineNitroConfig } from 'nitropack';

export default defineNitroConfig({
  preset: 'firebase',
});
```

**Reply by davidstackio** [reply]
I created an issue for this here https://github.com/unjs/nitro/issues/615

**Comment by DevJoghurt** [comment]
I have created a bugfix for the deployment problem: https://github.com/unjs/nitro/pull/627

In order to be able to create functions in addition to the nuxt app, I would currently suggest the following procedure:
### 1. Create a functions folder
The folder should be created next to the other nuxt folders in the root directory. Unfortunately firebase does not currently support proper monorepo setups. (See https://github.com/firebase/firebase-tools/issues/653)
### 2. Use the functions folder for firebase functions
In the function folder create a package.json and index.js with the required functions as described in the firebase docs.
### 3. IMPORTANT!!! Install dependencies separately
Since firebase does not yet support monorepos, dependencies must be installed separately.
`cd functions && npm install`
It is important to use npm here, as firebase cli looks for a package-lock.json.
### 4. Use the new codebase feature
Add the function folder in your firebase.json:
```json
{
  "functions": [{ 
    "source": "./.output/server", 
    "runtime": "nodejs16" ,
    "codebase": "nuxt"
  },{
    "source": "./functions",
    "runtime": "nodejs16",
    "codebase": "functions"
  }],
  "hosting": [
    {
      "site": "your-website",
      "public": "./.output/public",
      "cleanUrls": true,
      "rewrites": [
        { 
          "source": "**", 
          "function": "server" 
        }
      ]
    }
  ]
}
```
You can of course add more function folders if you have a use case for it.

When my bugfix is merged, it is then possible to deploy its app and all functions with the command `firebase deploy`, without any workaround or predeploy scripts.
